### PR TITLE
live callbacks for live-streams.

### PR DESCRIPTION
**Pull request by mhashim6**
With this, we can use continuous camera feed (eg, cv2::VideoCapture) and have live detection of scenes/events. Basically, issue #5

### CODE REVIEW

**Comment by mhashim6** [comment]
A usage example:
``` python
cam = cv2.VideoCapture(0)
stats_manager = StatsManager()
scene_manager = SceneManager()
scene_manager.add_detector(ContentDetector())
scene_manager.detect_scenes(frame_source=cam, callback=lambda frame: print("a new scene: ", len(frame)))
```

File: scenedetect/scene_manager.py [hunk 1]
```diff
@@ -491,13 +491,19 @@ def get_event_list(self, base_timecode):
...
 
 
-    def _process_frame(self, frame_num, frame_im):
+    def _process_frame(self, frame_num, frame_im, callback=None):
         # type(int, numpy.ndarray) -> None
         """ Adds any cuts detected with the current frame to the cutting list. """
         for detector in self._detector_list:
-            self._cutting_list += detector.process_frame(frame_num, frame_im)
+            cuts = detector.process_frame(frame_num, frame_im)
+            if len(cuts) and callback:
```
Comment (Breakthrough) [inline scenedetect/scene_manager.py]: Can this just be `if cuts and callback:`?

File: scenedetect/scene_manager.py [hunk 2]
```diff
@@ -491,13 +491,19 @@ def get_event_list(self, base_timecode):
...
+                callback(frame_im)
+            self._cutting_list += cuts
         for detector in self._sparse_detector_list:
-            self._event_list += detector.process_frame(frame_num, frame_im)
+            events = detector.process_frame(frame_num, frame_im)
+            if len(events) and callback:
```
Comment (Breakthrough) [inline scenedetect/scene_manager.py]: Same as above.  I think PyLint will flag this line is why.

File: scenedetect/scene_manager.py [hunk 3]
```diff
@@ -516,9 +522,9 @@ def _post_process(self, frame_num):
 
 
     def detect_scenes(self, frame_source, end_time=None, frame_skip=0,
-                      show_progress=True):
+                      show_progress=True, callback=None):
         # type: (VideoManager, Union[int, FrameTimecode],
-        #        Optional[Union[int, FrameTimecode]], Optional[bool]) -> int
+        #        Optional[Union[int, FrameTimecode]], Optional[bool], optional[callable]) -> int
```
Comment (Breakthrough) [inline scenedetect/scene_manager.py]: optional[callable] should be Optional[Callable[numpy.ndarray]]

File: scenedetect/scene_manager.py [hunk 4]
```diff
@@ -491,13 +491,19 @@ def get_event_list(self, base_timecode):
...
 
-    def _process_frame(self, frame_num, frame_im):
+    def _process_frame(self, frame_num, frame_im, callback=None):
         # type(int, numpy.ndarray) -> None
         """ Adds any cuts detected with the current frame to the cutting list. """
         for detector in self._detector_list:
-            self._cutting_list += detector.process_frame(frame_num, frame_im)
+            cuts = detector.process_frame(frame_num, frame_im)
+            if len(cuts) and callback:
+                callback(frame_im)
```
Comment (Breakthrough) [inline scenedetect/scene_manager.py]: Should the frame number or timecode be passed to the callback as well?

Comment (mhashim6) [reply scenedetect/scene_manager.py]: I can do the frame number, but where do I obtain the timecode?

**Comment by mhashim6** [comment]
> Looks good overall, suggest using PyLint or something similar though.
> 
> Very minor changes requested for code style/quality, but overall great PR!

Thanks! I installed PyLint in PyCharm and pushed some changes.

File: scenedetect/scene_manager.py (hunk 4 above)
Comment (mhashim6) [reply scenedetect/scene_manager.py:500]: @Breakthrough are you available?

Comment (Breakthrough) [reply scenedetect/scene_manager.py:500]: Good point, I forgot the scene manager only deals with frame numbers.  Sorry for the delay in getting back to you on this one, will definitely be trying to get this in for the next release though!

Comment (mhashim6) [reply scenedetect/scene_manager.py]: No problem at all. I pushed a commit with `frame_num`.

**Comment by Breakthrough** [comment]
I may retarget this for the v0.6.x branch before merging it, and have to update the documentation with some examples, but this will definitely make it in for the next release.  Thank you so much for this!

**Comment by mhashim6** [comment]
> I may retarget this for the v0.6.x branch before merging it, and have to update the documentation with some examples, but this will definitely make it in for the next release. Thank you so much for this!

This is great! You are welcome!
In fact, thank you in the first place, we used this lib + live feed feature in our [graduation project](https://www.linkedin.com/pulse/blind-people-dont-have-good-muhammad-hashim-1f/) : )

**Comment by Breakthrough** [comment]
@mhashim6 you are most welcome!  Would it be okay if I referenced your paper on the following page:
https://pyscenedetect.readthedocs.io/en/latest/other/resources/

(If not, no worries!)  Thank you so much!

**Comment by mhashim6** [comment]
> @mhashim6 you are most welcome!  Would it be okay if I referenced your paper on the following page:
> 
> https://pyscenedetect.readthedocs.io/en/latest/other/resources/
> 
> 
> 
> (If not, no worries!)  Thank you so much!

Oh not at all, I'd actually be glad! Thank you!

**Comment by research-boy** [comment]
> A usage example:
> 
> ```python
> cam = cv2.VideoCapture(0)
> scene_manager.add_detector(ContentDetector())
> scene_manager.detect_scenes(frame_source=cam, callback=lambda frame: print("a new scene: ", len(frame)))
> ```

when i tried this on a live video input, it throws error,
````
    scene_manager.detect_scenes(frame_source=cam, callback=lambda frame: print("a new scene: ", len(frame)))
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/scene_manager.py", line 752, in detect_scenes
    self._base_timecode = video.base_timecode
AttributeError: 'cv2.VideoCapture' object has no attribute 'base_timecode'
```

**Comment by Breakthrough** [comment]
@research-boy The API has changed slightly in v0.6, you now need to do something like:

    from scenedetect import VideoStreamCv2, SceneManager, ContentDetector
    scene_manager = SceneManager()
    cam = VideoStreamCv2(0)
    scene_manager.add_detector(ContentDetector())
    scene_manager.detect_scenes(video=cam, callback = ...)

Hope this helps!

**Comment by research-boy** [comment]
@Breakthrough Thanks for the example.
With the use case i am trying to get this run, the input is  from a gstreamer pipe.
Like:
```r
# in opencv how i read 
cap = cv2.VideoCapture("videotestsrc ! appsink")
# As per the example you shared
cam = VideoStreamCv2("videotestsrc ! appsink")
```
 when i tried the example with this it throws error
```r
    cam = VideoStreamCv2(source)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/backends/opencv.py", line 92, in __init__
    self._open_capture(framerate)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/backends/opencv.py", line 297, in _open_capture
    raise OSError('Video file not found.')

```

**Comment by Breakthrough** [comment]
Ah thanks for sharing, that definitely needs to be fixed.  This check needs to also see if there is a `!` in the path:
https://github.com/Breakthrough/PySceneDetect/blob/master/scenedetect/backends/opencv.py#L296=
As a workaround, you can fall back to using v0.5, or comment the above check out temporarily.

In the meantime, I've filed #276 and will make sure to include a fix for this in the next release (v0.6.1.).  Thanks for bringing this to my attention, and sorry for the inconvenience.

**Comment by Breakthrough** [comment]
@research-boy do you also need to set `cv2.CAP_GSTREAMER` when you are opening the `cv2.VideoCapture`?  In v0.6.1 I may instead add the ability for you to create a VideoStreamCv2 from an existing cv2.VideoCapture.

Thanks!

**Comment by research-boy** [comment]
> Ah thanks for sharing, that definitely needs to be fixed. This check needs to also see if there is a `!` in the path: https://github.com/Breakthrough/PySceneDetect/blob/master/scenedetect/backends/opencv.py#L296= As a workaround, you can fall back to using v0.5, or comment the above check out temporarily.
> 
> In the meantime, I've filed #276 and will make sure to include a fix for this in the next release (v0.6.1.). Thanks for bringing this to my attention, and sorry for the inconvenience.

Ya , the video feed is getting initialized  right now after commenting out those lines.

 

> @research-boy do you also need to set `cv2.CAP_GSTREAMER` when you are opening the `cv2.VideoCapture`? In v0.6.1 I may instead add the ability for you to create a VideoStreamCv2 from an existing cv2.VideoCapture.
> 
> Thanks!

cv2.CAP_GSTREAMER is not necessary , opencv by default recognizes the argument accordingly


After the video is initialized i am getting another error:
```
[ WARN:0] global /opencv-build/opencv/modules/videoio/src/cap_gstreamer.cpp (1081) open OpenCV | GStreamer warning: Cannot query video position: status=0, value=-1, duration=-1
[ WARN:0] global /opencv-build/opencv/modules/videoio/src/cap_gstreamer.cpp (1217) getProperty OpenCV | GStreamer warning: unhandled property: 6
Video codec detection failed. If output is incorrect:
  - Re-encode the input video with ffmpeg
  - Update OpenCV (pip install --upgrade opencv-python)
  - Use the PyAV backend (--backend pyav)
For details, see https://github.com/Breakthrough/PySceneDetect/issues/86
Traceback (most recent call last):
  File "tm_sd.py", line 11, in <module>
    scene_manager.detect_scenes(video=cam, callback=lambda frame: print("a new scene: ", len(frame)))
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/scene_manager.py", line 801, in detect_scenes
    self._process_frame(position.frame_num, frame_im, callback)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/scene_manager.py", line 676, in _process_frame
    cuts = detector.process_frame(frame_num, frame_im)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/detectors/content_detector.py", line 147, in process_frame
    frame_score = self._calculate_frame_score(frame_num, curr_hsv, last_hsv)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/detectors/content_detector.py", line 98, in _calculate_frame_score
    delta_h, delta_s, delta_v, delta_content = calculate_frame_score(curr_hsv, last_hsv)
  File "/usr/local/lib/python3.6/dist-packages/scenedetect/detectors/content_detector.py", line 48, in calculate_frame_score
    numpy.abs(current_frame_hsv[i] - last_frame_hsv[i])) / float(num_pixels)
ValueError: operands could not be broadcast together with shapes (1080,1920) (243,360) 
terminate called without an active exception
Aborted (core dumped)

```

**Comment by Breakthrough** [comment]
Would you be able to file a new bug report for this in the issue tracker?  I would like to get some more information from you about how to reproduce this, but would rather start a new issue specifically for this case.  Thanks!

If possible, in the bug report, could you include some steps for how I can reproduce the issue?  I am not familiar with gstreamer or pipes (do they work with the `opencv-python` package from `pip` or did you compile OpenCV yourself?), so if you could point me in the right direction for how I can reproduce the issue, that would be very helpful.

**Comment by Breakthrough** [comment]
Also as a workaround, you can use detectors directly with a VideoCapture object, e.g.:

```
import cv2
from scenedetect import ContentDetector

def my_callback(frame_num, frame_img):
    print("Found a scene at frame %d" % frame_num)

cap = cv2.VideoCapture("../PySceneDetect/tests/resources/goldeneye.mp4")
detector = ContentDetector()
frame_num = 0
while True:
    retval, frame_img = cap.read()
    if not retval:
        break
    frame_num += 1
    for frame_num in detector.process_frame(frame_num, frame_img):
        my_callback(frame_num, frame_img)
```

This might actually be a better way of integration in general for users needing live streams, as the main benefits to using a SceneManager come from processing an entire video, and the fact that it maintains a mapping of timecodes to each scene.  Tracking time when it comes to live streams/devices can be challenging, so this area of the API still needs a lot more thought.

**Comment by Breakthrough** [comment]
I've updated #276 as well for the next v0.6.1 release so this should work more smoothly by letting people pass their own `VideoCapture` objects using an adapter:
https://github.com/Breakthrough/PySceneDetect/blob/7913228/tests/test_api.py#L133-L153=
And also included various examples of this in the manual.

**Comment by fiosaint** [comment]
> Would you be able to file a new bug report for this in the issue tracker? I would like to get some more information from you about how to reproduce this, but would rather start a new issue specifically for this case (and keep the discussion there). Thanks!
> 
> If possible, in the bug report, could you include some steps for how I can reproduce the issue? I am not familiar with gstreamer or pipes (do they work with the `opencv-python` package from `pip` or did you compile OpenCV yourself?), so if you could point me in the right direction for how I can reproduce the issue, that would be very helpful.
> 
> It would also be useful to know if this issue exists on v0.5 as well, or if this is something that fails only in v0.6. I can certainly investigate further into that if you can point me in the right direction, or provide some steps for how to set this up.
> 
> Edit: Also, if I am reading the error correctly, it seems like the stream somehow returned two frames of different sizes. Are you using any kind of filters or anything which might change or crop the frame?

@Breakthrough  Thanks for the feedback.
As you mentioned the gstreamer pipe is returning a different size frame initially for  5 or 6 frames after that getting  original size which is 1080p .
I would like to skip those initial frames , would be great if you can guide me with this.

**Comment by research-boy** [comment]
> Do you know why those frames are smaller than the rest of the video, and is it always the same number of frames? Is there an easy way I can reproduce this on my end?
> 
> PySceneDetect requires each frame to be the same size, so you may want to skip over the first few frames until you start getting frames with the correct size. In v0.6.1 there will be a new adapter that lets you use a `cv2.VideoCapture` object directly: http://scenedetect.com/projects/Manual/en/v0.6.1/api/backends.html#devices-cameras-pipes
> 
> In the meantime, if you are still using a `VideoStreamCv2` object, you can call `read()` several times until it starts returning frames with the correct resolution, and _then_ call `detect_scenes`. This will effectively skip that part before doing the scene detection.
> 
> Hope this helps!

@Breakthrough as of now i am not sure why this is happening, i tried getting the frame directly in gstreamer itself , still it's same, it's look like that how the plugin is initializing the frame

**Pull request merged**